[FAB-17795] Build channel metadata retroactively if not present at peer start

wenjianqiao opened the pull request:
Signed-off-by: Wenjian Qiao <wenjianq@gmail.com>

#### Type of change
- New feature

#### Description
Channel metadata has been added to statecouchdb to store the mapping of db namespaces
to db names. However, the channel metadata is not available in v2.0/v2.1 peers. This PR
adds support to retroactively build channel metadata at peer start if such metadata
is not present in statecouchdb.

When a state CouchDB is opened (GetDBHandle), it retroactively builds channel metadata if
the metadata is not present. A NamespaceProvider is implemented to build possible namespaces for the channel. The possible namespaces will be verified by the existing databases and only namespaces matching existing databases will be added to channel metadata.

#### Additional details
The channel metadata will be used to remove per-channel databases and support checkpoint snapshots (e.g., create snapshot, clean up after bootstrap failure)

#### Related issues
https://jira.hyperledger.org/browse/FAB-17795
https://jira.hyperledger.org/browse/FAB-17787

wenjianqiao committed:
[FAB-17795] Build channel metadata retroactively if not present

Channel metadata has been added to statecouchdb to store the mapping of db namespaces
to db names. However, the channel metadata is not available in v2.0/v2.1 peers. This PR
adds support to retroactively build channel metadata at peer start if such metadata
is not present in statecouchdb.

When a state CouchDB is opened (GetDBHandle), it retoactively builds channel metadata if
the metadata is not present. A NamespaceProvider is implemented to build possible namespaces
for the channel. The possible namespaces will be verified by the existing databases and
only namespaces matching existing databases will be added to channel metadata.

Signed-off-by: Wenjian Qiao <wenjianq@gmail.com>

cendhu commented on core/ledger/kvledger/kv_ledger_provider.go:
```diff
@@ -333,7 +334,8 @@ func (p *Provider) open(ledgerID string) (ledger.PeerLedger, error) {
 	// Get the versioned database (state database) for a chain/ledger
-	db, err := p.dbProvider.GetDBHandle(ledgerID)
+	channelInfoProvider := &channelInfoProvider{ledgerID, blockStore}
+	db, err := p.dbProvider.GetDBHandle(ledgerID, channelInfoProvider)
```
Isn't it good to encapsulate `DeployedChaincodeInfoProvider` in the `channelInfoProvider` and pass it as a single argument to this `GetDBHandle()`? This is because of the following reason: we need to pass two dependencies that would be used by the `GetDBHandle()`. One is being passed in `NewDBProvider()` and another in `GetDBHandle()`. Keeping this at one place is good IMO. Further, chaincode info is part of channel info.

cendhu commented on core/ledger/kvledger/txmgmt/privacyenabledstate/namespace_helper.go:
```diff
@@ -0,0 +1,121 @@
+	}
+	itr := &resultsItr{ns: ns, dbItr: dbItr}
+	return itr, nil
+}
```
```suggestion
func (sqe *simpleQueryExecutor) GetStateRangeScanIterator(ns string, startKey string, endKey string) (commonledger.ResultsIterator, error) {
	dbItr, err := sqe.vdb.GetStateRangeScanIterator(ns, startKey, endKey)
	return dbItr.(commonledger.ResultsIterator), err
}
```
Wouldn't this work?

At least for the range query, this redirection is bit forced due to duplication of the `ResultIterator` interface. In the `statedb.go`, we have a `ResultIterator` that is used by the each statedb implementation. Then, in the `common/ledger/ledger_interface.go`, we have the same interface. Due to the mixed usage, we couldn't simply reuse the iterator present at the `VersionedDB`. 

To address this correctly, we need to refactor the existing code and remove the duplicate interfaces used for range query iterators.

cendhu commented on core/ledger/kvledger/txmgmt/privacyenabledstate/namespace_helper.go:
```diff
@@ -0,0 +1,121 @@
+// Close implements method in interface ledger.ResultsIterator
+func (itr *resultsItr) Close() {
+	itr.dbItr.Close()
+}
```
```suggestion
// Next implements method in interface ledger.ResultsIterator
func (itr *resultsItr) Next() (commonledger.QueryResult, error) {
	queryResult, err := itr.dbItr.Next()
	return queryResult.(commonledger.QueryResult), err
}
```
Wouldn't this work? Let me know if I am missing something.

Again the redirection is forced due to duplication of the `QueryResult` interface. In the `statedb.go`, we have a `QueryResult` that is used by the each statedb implementation. Then, in the `common/ledger/ledger_interface.go`, we have the same interface. Due to the mixed usage, we couldn't simply reuse the iterator present at the `VersionedDB`.

wenjianqiao replied:
Since statecouchdb returns *statedb.VersionedKV but the caller expects *queryresult.KV, the data type conversion is needed.

cendhu commented on core/ledger/kvledger/txmgmt/privacyenabledstate/db.go:
```diff
@@ -95,11 +105,25 @@ func (p *DBProvider) RegisterHealthChecker() error {
+
 	bookkeeper := p.bookkeepingProvider.GetDBHandle(id, bookkeeping.MetadataPresenceIndicator)
 	metadataHint := newMetadataHint(bookkeeper)
 	return NewDB(vdb, id, metadataHint)
```
To reduce the extra code indentation. 

```suggestion
	bookkeeper := p.bookkeepingProvider.GetDBHandle(id, bookkeeping.MetadataPresenceIndicator)
	metadataHint := newMetadataHint(bookkeeper)
	
	db, ok := vdb.(statedb.ChannelMetadataCreator)
	if !ok || db.ChannelMetadataExists() {
	        	return NewDB(vdb, id, metadataHint)
	}
	logger.Warnf("No channel metadata found for channel %s, creating it now...", id)
	namespaces, err := constructStatedbNamespaces(id, vdb, channelconfigProvider, p.deployedccInfoProvider)
	if err != nil {
		return nil, err
	}
	if err = db.CreateChannelMetadata(namespaces); err != nil {
		return nil, err
	}
	return NewDB(vdb, id, metadataHint)
```

cendhu commented on core/ledger/kvledger/txmgmt/privacyenabledstate/namespace_helper.go:
```diff
@@ -0,0 +1,121 @@
+	retNamespaces = append(retNamespaces, "")
+	for _, ns := range deployedccInfoProvider.Namespaces() {
+		retNamespaces = append(retNamespaces, ns)
+	}
```
```suggestion
	retNamespaces = append(retNamespaces, deployedccInfoProvider.Namespaces()...)
```

cendhu commented on core/ledger/kvledger/txmgmt/privacyenabledstate/namespace_helper.go:
```diff
@@ -0,0 +1,121 @@
+	implicitCollNames := deployedccInfoProvider.ImplicitCollectionNames(mspids)
+	for _, implicitCollName := range implicitCollNames {
+		retNamespaces = append(retNamespaces, derivePvtDataNs("_lifecycle", implicitCollName))
+		retNamespaces = append(retNamespaces, deriveHashedDataNs("_lifecycle", implicitCollName))
```
why the namespace is `_lifecycle`? Don't we have an implicit collection per chaincode also? Should it be the `chaincodeName` which would include `_lifecycle` too?

For `_lifecycle` and each user chaincode, we would have one implicit collection. Need to verify for the `lscc` namespace.

wenjianqiao replied:
Thank you for pointing it out. Fixed.

cendhu commented on core/ledger/kvledger/txmgmt/privacyenabledstate/namespace_helper.go:
```diff
@@ -0,0 +1,121 @@
+	}
+
+	// add namespaces for chaincodes and explicit collections
+	chaincodesInfo, err := deployedccInfoProvider.AllChaincodesInfo(channelName, &simpleQueryExecutor{vdb})
```
The order has to be
1. Fetch namespaces of lifecycle management components -- add to namespace list
2. Fetch all chaincodes' name -- -- add to namespace list
3. Fetch the implicit collection name (without the chaincode prefix)
4. For each entry in the namespace list, construct and add the final implicit collection name for that namespace.

cendhu commented on core/ledger/kvledger/txmgmt/privacyenabledstate/namespace_helper.go:
```diff
@@ -0,0 +1,121 @@
+}
+
+func (sqe *simpleQueryExecutor) GetState(ns string, key string) ([]byte, error) {
+	versionedValue, err := sqe.vdb.GetState(ns, key)
```
No related to this PR. I understand that we implement the `ledger.SimpleQueryExecutor` here. The interface returns `VersionedValue` for the range query and only the `value` for the simple get. If we had kept this consistent, we could have simply passed the `VersionedDB` as `ledger.SimpleQueryExecutor`.

cendhu commented on core/ledger/kvledger/txmgmt/statedb/statecouchdb/statecouchdb.go:
```diff
@@ -261,14 +264,22 @@ func (vdb *VersionedDB) getNamespaceDBHandle(namespace string) (*couchDatabase,
+		// Therefore, the code path gets here when vdb.channelMetadata is still nil (i.e, before
+		// CreateChannelMetadata is called).
+		// It is safe to skip the namespace info for now because it will be added by CreateChannelMetadata later.
+		if vdb.channelMetadata != nil {
```
The problem occurs due to the changes made in the `newVersionedDB()`. if that function creates the `channelMetadata` when it is nil irrespective of the `savepoint` value, this wouldn't occur but it requires changes to `ChannelMetadataExists()`. IMO, this comment might create confusion later and see whether we can avoid this by retaining the old `newVersionedDB()`.

cendhu commented on core/ledger/kvledger/txmgmt/statedb/statecouchdb/statecouchdb.go:
```diff
@@ -837,6 +848,32 @@ func (vdb *VersionedDB) GetFullScanIterator(skipNamespace func(string) bool) (st
 
+// ChannelMetadataExists returns true if channel metadata already exists in VersionedDB
+func (vdb *VersionedDB) ChannelMetadataExists() bool {
+	return vdb.channelMetadata != nil
```
This is also related to the comment I made for the line 267 to 274. 

shouldn't it check for the savepoint? 

`ChannelMetadataExists()` to `IsChannelMetadataRecent()`

1. `savepoint` is **nil** -- a new channel and hence, it does not require any updates -- return true
2. `savepoint` is **not nil** -- not a new channel. If required, we can then check whether the channelMetadata has any mapping. If it does not have any, we can return false. Otherwise, without checking channelMetadata also, we can return false but needs to do extra processing at `CreateChannelMetadata()`

I do not have a strong opinion on the this. Hence, I leave it to you.

cendhu commented on core/ledger/kvledger/txmgmt/statedb/statecouchdb/statecouchdb.go:
```diff
@@ -837,6 +848,32 @@ func (vdb *VersionedDB) GetFullScanIterator(skipNamespace func(string) bool) (st
+		ChannelName:      vdb.chainName,
+		NamespaceDBsInfo: dbsInfo,
+	}
+	return vdb.writeChannelMetadata()
```
Shouldn't we check for the database existence and filter certain namespaces?

By default, we create a database for a chaincode or collection only when it is accessed first (applicable for implicit collection too). Hence, if we create mapping for all received namespaces and write the channelMetadata, it would end up an inconsistent data. To avoid this inconsistency, we need to check whether a db exist for a given namespace and only then add it to the channel metadata.

cendhu reviewed (commented):
Thanks, Wenjian. I have suggestions related to the overall approach.

wenjianqiao commented:
@cendhu Thank you for your comments. As discussed in our scrum, I have refactored the code and moved channelMetadata rebuild to newVersionedDB. I have addressed your comments unless the code that has been removed due to refactoring. 
The SimpleQueryExecutor and results iterator wrapper structs are remained the same because it requires refactoring.

cendhu commented on core/ledger/kvledger/txmgmt/privacyenabledstate/namespace_provider.go:
```diff
@@ -0,0 +1,139 @@
+		retNamespaces = append(retNamespaces, ccName)
+		for _, implicitCollName := range implicitCollNames {
+			retNamespaces = append(retNamespaces, derivePvtDataNs(ccName, implicitCollName))
+			retNamespaces = append(retNamespaces, deriveHashedDataNs(ccName, implicitCollName))
```
We need

1. hashed namespace for each implicit collection
2. private data namespace only for the peer's implicit collection

this code constructs the hashed and private data namespace for all implicit collections. As we have an additional checking at the `statecouchdb` pkg to ensure consistency, this would not create any issue but it could provide misinformation when someone reads the code. The misinformation could be that the private data namespace is created for all implicit collections (which is false) and hence, it creates confusion. 
The problem is also applicable for the explicit collections. Only the members of a collection would have a private data namespace. 

I understand that this code is temporary one to just fix a missing information in ledger and we will remove the code soon. Hence, I am okay with as-is if it requires more code to fix above confusion. For now, you can add a comment to clarify this scenario.

wenjianqiao replied:
@cendhu thanks for the comment. Initially I thought it would require new dependency to check peer membership for a collection, but just found that ledger already had MembershipProvider interface. Will add code to check eligibility before adding private data implicit/explicit collections.

cendhu replied:
For an implicit collection, this might work. For explicit collections, we only fetch the recent collection config using `DeployedCCInfoProvider.AllChaincodeInfo`. To find all members (including the past members) of a collections, we need to get all collection config from the collection config store. I am not sure whether this complexity is needed. Hence, I am okay if we can add a comment to avoid this confusion (as it is a temporary code).

cendhu commented on core/ledger/kvledger/txmgmt/privacyenabledstate/namespace_provider.go:
```diff
@@ -0,0 +1,139 @@
+
+	// iterate implicit collections, add namespaces for new lifecycle management namespace
+	for _, implicitCollName := range implicitCollNames {
+		retNamespaces = append(retNamespaces, derivePvtDataNs("_lifecycle", implicitCollName))
```
Same as the previous comment.

cendhu commented on core/ledger/kvledger/txmgmt/statedb/statecouchdb/couchdb.go:
```diff
@@ -459,6 +459,38 @@ func (couchInstance *couchInstance) retrieveApplicationDBNames() ([]string, erro
+// Note that false matching is possible if the channel name exceeds chainNameAllowedLength
+// because another channel may be truncated to the same name. Therefore, this method may
+// may return more than the channel's DB names.
+func (couchInstance *couchInstance) retrieveDBNamesMatchedForChannel(channelName string) (map[string]struct{}, error) {
```
Is this function needed?

Instead of allowing the false matching while trying to approximately fetch DBs associated with a channel, can't we simply use `retrieveApplicationDBNames()` at the `initChannelMetadata()` and remove this function. The logic in `initChannelMetadata()` takes care of finding the correct mapping between the namespace and db. This additional complexity does not seem necessary.

wenjianqiao replied:
Since we need to build a map to look up DBNames, this function builds the map with minimum entries to make dbname lookup more efficient. However, because this is one time call for v2.0/2.0 peers only, I agree we don't need the additional complexity.

cendhu commented on core/ledger/kvledger/txmgmt/statedb/statecouchdb/statecouchdb.go:
```diff
@@ -228,6 +215,10 @@ func newVersionedDB(couchInstance *couchInstance, redoLogger *redoLogger, dbName
 		return nil, err
 	}
 
+	if err = vdb.initChannelMetadata(savepoint == nil, nsProvider); err != nil {
```
To improve the readability, we can have

```
isNewDB := savepoint == nil
if err = vdb.initChannelMetadata(isNewDB, nsProvider)
```

This helps in providing a context for `savepoint == nil`.

cendhu reviewed (commented):
Thanks, Wenjian, for the changes. I have a few more suggestions. Once they are addressed, I will look at the test.

manish-sethi commented on core/ledger/kvledger/kv_ledger_provider.go:
```diff
@@ -333,7 +333,9 @@ func (p *Provider) open(ledgerID string) (ledger.PeerLedger, error) {
 	// Get the versioned database (state database) for a chain/ledger
-	db, err := p.dbProvider.GetDBHandle(ledgerID)
+	channelInfoProvider := &channelInfoProvider{ledgerID, blockStore}
+	namespaceProvider := privacyenabledstate.NewNamespaceProvider(channelInfoProvider, p.initializer.DeployedChaincodeInfoProvider)
```
This pattern seems a little odd to me. Here you construct a dependency `namespaceProvider` from a package and supply back to the code in the same package. I would rather pass `channelnfoProvider` as a dependency straight to the privacyenabledstate package.

manish-sethi commented on core/ledger/kvledger/txmgmt/statedb/statedb.go:
```diff
@@ -77,6 +78,12 @@ type VersionedDB interface {
+type NamespaceProvider interface {
+	GetStateDBNamespaces(channelName string, vdb VersionedDB) ([]string, error)
+}
+
```
Nothing wrong with this but for a single function interface, particularly when it is not intended to be embedded in another interface, I find it easier to define a function instead. You don't need to generate mock for testing as well - an [example](https://github.com/hyperledger/fabric/blob/3d38abe0043d1aa4727e301c517279097a45ce55/common/ledger/snapshot/file.go#L21). However, as I said that it's a matter of preference and want to leave up to you.

wenjianqiao replied:
Thank you @manish-sethi. Agree defining a function may be simpler in some cases. However, will keep the interface for this one, because we need to mock it multiple times in test code, and it is easier to mock an interface.

manish-sethi commented on core/ledger/kvledger/txmgmt/privacyenabledstate/namespace_provider.go:
```diff
@@ -0,0 +1,146 @@
+			retNamespaces = append(retNamespaces, deriveHashedDataNs(ccInfo.Name, implicitCollName))
+			retNamespaces = append(retNamespaces, derivePvtDataNs(ccInfo.Name, implicitCollName))
+		}
+		retNamespaces = append(retNamespaces, deriveHashedDataNs("_lifecycle", implicitCollName))
```
Rather than hard coding "_lifecycle", I would prefer to use `p.Namespaces()`. I assume that you did this for excluding the `lscc`. Though, it does not harm to include `lscc` as well still, we you want to remove `lscc`, I would prefer to remove `lscc` explicitly from `p.Namespaces()`. When we will remove `lscc` entirely, we would remove that hardcoding here as well.

manish-sethi commented on core/ledger/kvledger/txmgmt/privacyenabledstate/namespace_provider.go:
```diff
@@ -0,0 +1,146 @@
+
+// simpleQueryExecutor implements ledger.SimpleQueryExecutor interface
+type simpleQueryExecutor struct {
+	vdb statedb.VersionedDB
```
if you embed `statedb.VersionedDB`, you don't need to implement first and third functions below.

wenjianqiao replied:
Changed to embed statedb.VersionedDB. But 1st and 3rd functions are needed because they overwrite the same functions in VersionedDB.

manish-sethi commented on core/ledger/kvledger/txmgmt/privacyenabledstate/namespace_provider.go:
```diff
@@ -0,0 +1,146 @@
+// the collection config store. Because this is a temporary function needed to retroactively build namespaces
+// when upgrading v2.0/2.1 peers to a newer v2.x version and because returning extra private data namespaces
+// does not cause inconsistence, it makes sense to use the simple implementation.
+func (p *NamespaceProvider) GetStateDBNamespaces(channelName string, vdb statedb.VersionedDB) ([]string, error) {
```
In the connection with the comment in `kv_ledger_provider`, after looking at this code here, I feel that the logic in this funciton better suits in the package `kvledger`. That would prevent this high level logic to percolate much deeper in the code - wdyt?

cendhu replied:
Due to the dependency on `deriveHashedDataNs()` and `derivePvtDataNs()`, it may be bit challenging to move the whole `GetStateDBNamespaces()` to `kvledger` pkg.

wenjianqiao replied:
This function has logic (e.g. derivePvtDataNs) specific to the `privacyenabledstate` package, therefore cannot be moved to kvledger. Will split it into two functions: one is moved to channelInfoProvider in the `kvledger` package and the other stays in `privacyenabledstate` package.

manish-sethi reviewed (commented):
The logic looks good to me. However, I have a couple comments, mostly related to reorganizing the code a bit for a better separation of concerns.

manish-sethi commented on core/ledger/kvledger/txmgmt/privacyenabledstate/db.go:
```diff
@@ -429,3 +429,41 @@ func getIndexInfo(indexPath string) *indexInfo {
 	return indexInfo
 }
+
+// channelInfoProvider interface enables the privateenabledstate package to retrieve all the config blocks
+// and  namespaces and collections.
```
forgot to update the comments?
I though that at this level, perhaps you would name this a bit more restrictive like `nsCollNames` or `nsCollNamesProvider` instead of a generic name `channelInfoProvider`

manish-sethi commented on core/ledger/kvledger/txmgmt/privacyenabledstate/db_test.go:
```diff
@@ -587,3 +588,50 @@ func generateLedgerID(t *testing.T) string {
+
+//go:generate counterfeiter -o mock/channelinfo_provider.go -fake-name ChannelInfoProvider . channelInfoProviderWrapper
+
+// define this interface to break circular dependency
+type channelInfoProviderWrapper interface {
+	channelInfoProvider
+}
```
not sure whether this wrapper is needed, as the actual interface itself is unexported.

manish-sethi commented on core/ledger/kvledger/txmgmt/privacyenabledstate/db.go:
```diff
@@ -429,3 +429,41 @@ func getIndexInfo(indexPath string) *indexInfo {
+// and  namespaces and collections.
+type channelInfoProvider interface {
+	// NamespacesAndCollections returns namespaces and collections for the channel.
+	NamespacesAndCollections(vdb statedb.VersionedDB) (map[string][]string, error)
```
The earlier division was better where the query executor was still in this package and this function could the queryexector but it's fine for now.

manish-sethi commented on core/ledger/kvledger/txmgmt/statedb/statecouchdb/statecouchdb_test.go:
```diff
@@ -1479,6 +1481,81 @@ func TestChannelMetadata_NegativeTests(t *testing.T) {
+	}
+
+	// test an existing DB with channelMetadata, namespace provider should not be called
+	require.NoError(t, vdb.initChannelMetadata(false, fakeNsProvider))
```
a test with "true" flag could also be added.

manish-sethi reviewed (approved):
Minor comments, if you agree and want to address in a separate PR.
